parsimonious/grammar.py

```diff
@@ -448,9 +466,18 @@ def visit_rules(self, node, rules_list):
         rule_map.update(self.custom_rules)
 
         # Resolve references. This tolerates forward references.
-        done = set()
-        rule_map = OrderedDict((expr.name, self._resolve_refs(rule_map, expr, done))
-                               for expr in itervalues(rule_map))
+        # We use a job pool `to_resolve` to remember all rules to resolve. It is
+        # initialized with all existing rules in `rule_map` and all
+        # LazyReference rules found later will be added as well.
+        to_resolve = set(itervalues(rule_map))
```

Comment:
`set(rule_map.values())` is more pythonic and basically the same speed for an `OrderedDict` based on my benchmarking:

## Python 2.7
```
In [13]: from collections import OrderedDict

In [14]: dct = OrderedDict((i, i) for i in range(1000))

In [15]: %timeit set(itervalues(dct))
1000 loops, best of 3: 192 µs per loop

In [16]: %timeit set(viewvalues(dct))
1000 loops, best of 3: 229 µs per loop

In [17]: %timeit set(dct.values())
10000 loops, best of 3: 193 µs per loop
```

## Python 3.7
```
In [6]: from collections import OrderedDict                                                                                                               

In [7]: dct = OrderedDict((i, i) for i in range(1000))                                                                                                    

In [8]: %timeit set(itervalues(dct))                                                                                                                      
36 µs ± 275 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [9]: %timeit set(viewvalues(dct))                                                                                                                      
36 µs ± 462 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [10]: %timeit set(dct.values())                                                                                                                        
35.7 µs ± 315 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Looking at the implementation in the collections module, this is presumably because the runtime is dominated by key lookups which happen in both `OrderedDict.values` and `ValuesView.__iter__`.

Reply:
I agree. itervalues construct was used by the original code and I try not to remove it in afraid of breaking the py2/3 compatibility, and not to distract from the discussion of the algorithm. But given you verified it, I would prefer to use values() instead, too.